src/charm.py

```diff
@@ -142,21 +175,30 @@ def param_get(p):
             return str(event.params.get(p, ""))
 
         # Param order matters here because test.sh uses $1, $2, etc.
-        args = [param_get(param) for param in ["focus", "skip", "parallelism", "timeout", "extra"]]
+        args = [param_get(param) for param in ["focus", "skip", "parallelism", "timeout"]]
+        args += shlex.split(param_get("extra"))
+
         command = ["scripts/test.sh", *args]
 
         if not self._check_kube_config_exists(event):
             return
 
-        event.log(f"Running this command: {' '.join(command)}")
+        logger.info("Running scripts/test.sh: %s", "".join(command))
+
+        self.unit.status = MaintenanceStatus("Tests running...")
 
-        process = subprocess.run(command, capture_output=False, check=True)
+        # Let check=False so the log and process return code can be checked below.
+        process = subprocess.run(command, capture_output=False, check=False)
+
+        previous_status = self.unit.status
 
         if self._log_has_errors(event) or process.returncode != 0:
-            sys.exit(process.returncode)
+            event.fail("One or more tests failed.")
         else:
             event.set_results({"result": "Tests ran successfully."})
```

Comment:
I'm ok with this -- but here's the difference
The action does not fail with the same return code as before

for example)
previously the action might fail with a return code of 124

```sh
$ > juju run kubernetes-e2e/0 test
....
failure
...
$ > echo $?
124
```

but this change drops the `sys.exit(process.returncode)`

```sh
$ > juju run kubernetes-e2e/0 test
....
failure
...
$ > echo $?
-1?  <- generated by ops/juju with the event.fail
```


i don't know if ANYONE cares about this -- but it's a subtle UX change

Reply:
Good to know. I think it's okay to leave it as event.fail which allows us to show a message to justify the failure in `juju show task`.

```
❯ juju show-task 70 | head
Action id 70 failed: One or more tests failed.
junit: /home/ubuntu/70-junit.tar.gz
log: /home/ubuntu/70.log.tar.gz
```